### ISSUE TITLE
[Feature request] Unsplit-block change type?

### ISSUE THREAD

**Issue by varunarora** [facebook/draft-js#149]
Not sure is this is the right place, but my apologies if issues is strictly for bugs.

When adding a new block (or splitting an existing block), the `onChange` event appropriately reports the `lastChangeType` on the `EditorState` object to be `split-block`. But the reverse action, of using backspace to unsplit it or reduce the extra line just gives us the standard `backspace-character`. Would it possible to make a more meaningful change type returned?

My use case is the following: I want to extract content from a series of continuous blocks of the same type, or split blocks (in Draft JS terms). I couldn't find an existing helper function so I made my own simple helper `getContinuousBlocksOfType` (which I think would be hugely helpful to have inside the core of draftjs) which just builds a map of all the split blocks (an array) originating from the original block key. Building this array forward is easy because I can watch for `split-block` change types, but I would like not to keep watching for `backspace-character` when a block has been eliminated.

**Comment by hellendag** [comment]
> Not sure is this is the right place, but my apologies if issues is strictly for bugs.

Issues are fine for other support as well. :)

Can you provide more details about your use case? What kind of editor is it?

**Comment by varunarora** [comment]
It is one gigantic full-page `Editor` component with multiple sections which open with section headings (think entire form in one `contentEditable` input). The purpose is to help teachers make lesson plans, with fluid sections; they should be able to change the titles, remove sections, move them around. (Will most likely put Draft in production in a month for this)

At the moment, each section heading is a block, followed a block for its contents. So two blocks per section. I want to constantly keep reading the "contents" of the section (here, I mean the second of the two blocks) being edited, so the app can do interesting inferences on the it. I realized that to do this, I have to constantly monitor the splits that happen to that block so I can read the collective input of all paragraphs in that section.

On a sidenote, do you, by any chance, suggest that instead of using one big component, I should have several components for the same use case?

**Comment by hellendag** [comment]
> On a sidenote, do you, by any chance, suggest that instead of using one big component, I should have several components for the same use case?

It depends on your intended UI. I think that for large forms, it often makes sense to break things apart. The major tradeoff is that you can't select-all on the entire document, and you would have to implement key behavior to move the cursor from one editor to the next. If that's not so important, then separate editor instances could make a lot of sense. (There may be other drawbacks I'm not thinking of, but certainly the application simplicity would be a plus.)

I think this would also allow you to more easily observe changes within individual sections.

In fact, you could go a step further and use separate editors for the headers as well. In Notes, for example, the title editor is a separate Draft instance. The user may tab from that field into the main editor.

If you do want to go the route of continuing to use a single editor, one thing you might do is track your sections independent of the `EditorState`. For example, let's say you have a document like this:

```javascript
blocks: [
  // first section starts here
  {
    key: 'foo',
    text: 'Preamble',
    type: 'header-one',
  },
  {
    key: 'bar',
    text: 'We the People of the United States',
    type: 'unstyled',
  },
  // second section starts here
  {
    key: 'baz',
    text: 'Second Section',
    type: 'header-one',
  },
  {
    key: 'derp',
    text: 'Provide for the common defence, promote the general Welfare',
    type: 'unstyled',
  },
]
```

At the top level, you could track which blocks belong together in sections:

```javascript
sections: {
  preamble: ['foo', 'bar'],
  second: ['baz', 'derp'],
},
```

It means you would have to evaluate your `ContentState` on every change to determine the sections, but it might give you a way to observe per-section changes.

I'm not sure if I really *recommend* this approach, but it's an idea.

**Comment by varunarora** [comment]
@hellendag, you are amazing. I had NO IDEA I could choose my own keys to begin with. I was doing all of this except for specifying the `key`s on the raw blocks. So wait - when `split-blocks` happens on blocks whose keys I have already specified for trackability, do I get an array back on `getBlockForKey`? That would be PERFECT.

Thanks for your other thoughts! I'd have to observe user behavior to see if splitting them into multiple components makes sense or not. For now, I'll just roll with this, lower-proficient users just like MS-Word emulations with things like select-all.

**Comment by varunarora** [comment]
Na, it doesn't do that :( My problem is still as bad; tracking all split blocks.

**Comment by victorhqc** [comment]
@hellendag I'm interested in that solution you propose with two different Draft instances. The thing is I need to handle the return event in write and paste. Still new to this, if there's an example somewhere it would be really useful :)

**Comment by victorhqc** [comment]
Well, seeing how Facebook notes look, multi-line title isn't so bad :)

**Comment by victorhqc** [comment]
Well, seeing how Facebook notes look, multi-line title isn't so bad :)

**Comment by tobiasandersen** [comment]
I also think this could be useful. Here's why:

Each time a new block is created, we need to connect that block, using the block's key, to a new instance of a mobx store. It basically works the same way entities does (before #376). We currently do this in `onChange` when `lastChangeType === 'split-block'`, and it works fine. 

But it's a bit more cumbersome to remove a store instance when the block that it's connected to is removed. Here're two solutions I could think of:

1. check for `lastChangeType === 'backspace'` and see if the currently selected block is empty
2. count the blocks for the old editorState and the new and see if there's a difference.

While it's definitely not a huge pain, I think it'd be a more unified solution to check for added blocks using `split-block`, and for removed blocks using something like `remove-block`. If not, maybe this issue could be closed?

**Comment by mitermayer** [comment]
Closing this thread for now, but if there is still interest in this discussion please feel free to reopen/comment or create a new issue around it.

**Comment by varunarora** [comment]
@mitermayer Very much still interested in discussing a solution :)

**Comment by JonathanPrada** [comment]
I'm also interested in a solution for this. At the moment It feels like embedding an editor within each section component might be more simple.
